src/parser.rs

```diff
@@ -201,6 +201,19 @@ impl<'i> Input<'i> {
         Input::with_log(input, None)
     }
 
+    pub fn no_trim(input: &'i str) -> Self {
+        Input {
+            chars: input.chars(),
+        }
+    }
+
+    pub fn trim_tab_and_newlines(original_input: &'i str) -> Self {
+        let input = original_input.trim_matches(ascii_tab_or_new_line);
```

Comment:
Mmmh, when trimming those, are they no longer syntax violations?

Reply:
Oh my bad, they are !

Reply:
I don't understand anymore what this code is supposed to be doing anymore. Why are we not trimming `c0_control_or_space` anymore? This method looks oddly like `with_log` now.

Reply:
As defined [in the basic url parser spec](https://url.spec.whatwg.org/#concept-basic-url-parser), removing C0 control or space from input only occurs when If url is not given.

Removing tabs and newlines occurs either way though.

This means we need to use different `parser::Input` s.

I think It would be more explicit if with_log / trim_tab_and_newlines mentionned that, ie 
parser::Input::with_base_url / parser::Input::no_base_url or something, or having a factory that would decide on the best parser to build depending on having a base_url.

Setters would use the former, and the default parser would use the latter.
What do you think ?